### PR TITLE
[THREESCALE-1517] Retry policy

### PR DESCRIPTION
**Pull request by davidor**
Fixes [THREESCALE-1517] (https://issues.jboss.org/browse/THREESCALE-1517)

This PR adds the Retry policy. The PR contains a README that explains how the policy works and how to configure it.

### CODE REVIEW

**Comment by davidor** [comment]
@eloycoto I added a couple of test cases (tests 7 and 8) to verify that everything works when `APICAST_UPSTREAM_RETRY_CASES` is empty.

File: gateway/conf.d/apicast.conf [hunk 1]
```diff
@@ -115,6 +115,18 @@ location @upstream {
   #   {% endif %}
   # {% endif %}
 
+  # When 'upstream_retry_cases' is empty, apply the same default as NGINX.
+  # If the proxy_next_upstream directive is not declare, the retry policy
```
Comment (dfennessy) [inline gateway/conf.d/apicast.conf]:
```suggestion
  # If the proxy_next_upstream directive is not declared, the retry policy
```

Comment (davidor) [reply gateway/conf.d/apicast.conf]: Fixed. Thanks for the review @dfennessy :+1:

File: gateway/src/apicast/balancer.lua [hunk 2]
```diff
@@ -70,6 +70,24 @@ local function set_timeouts(balancer, timeouts)
   end
 end
 
+local function set_balancer_retry(balancer)
+  local _, err = balancer:set_more_tries(1)
```
Comment (mikz) [inline gateway/src/apicast/balancer.lua]: Wouldn't it be nicer as `balancer:retry_next_request()` ?  Wrapping the OpenResty API into something more human-friendly.

Comment (davidor) [reply gateway/src/apicast/balancer.lua]: Good idea :+1:
Changed.

**Pull request merged**